server/pbenchinacan/load_keycloak.sh

```diff
@@ -91,7 +91,7 @@ curl -si -f -X POST "${KEYCLOAK_HOST_PORT}/admin/realms/${REALM}/client-scopes"
           "protocolMapper": "oidc-audience-mapper",
           "consentRequired": false,
           "config": {
-            "included.client.audience": "pbench-dashboard",
+            "included.client.audience": "'${CLIENT_ID}'",
```

Comment:
Do you really want the single quotes here? That is, before we had `pbench-dashboard` and now we have `'pbench-dashboard'` ... one of these might be wrong?? Even if Keycloak de-quotes the audience, if the old form worked then the quotes in the new form aren't necessary.

Reply:
Single quotes here act as escape characters, the entire `-d` (data) value of a request is bounded by single quotes so I had to use the single quotes here otherwise it treats `${CLIENT}` as a literal string.

Reply:
@npalaska, you really should be careful with your terminology:  single quotes are no "escape characters", here or anywhere else.  The single quotes in line 94 are _ending_ on quoted section and _beginning_ another, leaving the shell-variable-expansion **outside** the quoted expression.

And, in Dave's defense, this approach makes it very easy to misunderstand what is happening.

Instead of using single quotes around the entire expression, we should consider a different approach, such as building up a variable with the argument in it and then quoting it once where it is expanded, or using a HERE document, or something.

Reply:
We could try something like this:
```
curl -si -f -X POST "${KEYCLOAK_HOST_PORT}/admin/realms/${REALM}/client-scopes" \
  -H "Authorization: Bearer ${ADMIN_TOKEN}" \
  -H "Content-Type: application/json" \
  -d - <<__HERE__
    {
      "name": "pbench",
      "description": "",
      "protocol": "openid-connect",
      "attributes": {
        "include.in.token.scope": "true",
        "display.on.consent.screen": "true",
        "gui.order": "",
        "consent.screen.text": ""
      },
      "protocolMappers": [
        {
          "name": "pbench-mapper",
          "protocol": "openid-connect",
          "protocolMapper": "oidc-audience-mapper",
          "consentRequired": false,
          "config": {
            "included.client.audience": "${CLIENT_ID}",
            "id.token.claim": "false",
            "access.token.claim": "true"
          }
        }
      ]
    }
__HERE__
```

Reply:
I tried this but the curl request results in `400-BAD_REQUEST` with this message `Mark bundle as not supporting multiuse`. Tried looking around for what `multiuse` means but I didn't find much, instead found something on [multiplexing](https://blog.cloudflare.com/http-2-for-web-developers/) which I am not sure is related. But again didn't spend too much time fixing this so I kept it as it is now.

Reply:
OK, well, we just need to remember to be suspicious any time we see two different kinds of quotation marks in the same line of shell script code....

(Thanks for giving it a try -- I, too, had problems with `curl -d` in the script which sets container image tag expirations -- I couldn't get it to work without specifying a content-length header, which I thought was awkward and silly, but it probably has something to do with the fact that it's actually reading from a pipe and pipes don't have explicit lengths.)